### ISSUE TITLE
DISCUSSION: How does this differ from "Composable Model" approach?

### ISSUE THREAD

**Issue by james-shaw-turner** [dchambers/composerjs#7]
There are three core differences between ComposerJS and the pre-existing "Composable Model" (CM) library. 
1. ComposerJS handlers are called with input and output properties that it updates. ComposerJs is responsible for calling dependent handlers with the input values derived from previous handlers. CM handlers are  passed a model sandbox and registers event listeners.  It sets field values directly - that trigger cascaded events. See https://github.com/dchambers/composerjs/issues/13 for detailed pros and cons
2. ComposerJS has handler functions, CM has handler objects.
3. ComposerJS fixes the shape of the entire object tree including sub-nodes on sealing. CM only fixes the shape of the fields of each node. 
#1. General discussion of property chaining

In both proposals a handler may be  dependent on a field/property that is updated by another handler. In the general case property chaining is represented by a Directed Acyclic Graph (DAG).  Both proposals must cope with any arbitrary DAG.
## Composable Model (CM) approach

The CM registers handler **objects**. The handler is passed a sandboxed instance of the model in its `init()` method. The handler then adds listeners to any field it wants to use as an input and calls `set(name, value)` to update its output fields. When a property is set externally or internally (by a handler) any interested handler listeners methods are invoked - chaining follows automatically. 
## ComposerJS approach

ComposerJS registers handler **functions**. The function is passed model **values** as input and output parameters. The handler function is called whenever one of the inputs changes.
Whenever a property is updated externally the model must find any dependent handlers and execute those. It must then iteratively call any other dependent handlers - ensuring all handlers are executed in the correct order. 
The current example only shows a chain starting with an external `set(name, value)` method. It is not clear how it works when handlers are the source of updates. 
#2. Object versus function handlers
## CM has handler objects

CM handlers are objects that implement the handler interface. As such they have two major life-cycle methods: `init()` which is called immediately after sealing, and `destory()`. This allows a handler to create any resources based on the fixed initial values of the model. e.g. a price rate provider must create a subscription that depends on currency-pair and amount. Similarly when the model is disposed of the `destroy()` method allows the subscription to be freed. 
Handlers may have "business semantic" methods that can be executed directly on the handler or exported onto the model object. e.g. `model.executeTrade("buy")`; 
## ComposerJS has handler functions

Although ComposerJS is passed a handler function, in JavaScript functions are first class objects so we can add properties as we want. Also the function can just be a method of a containing object. So this is just a stylistic, rather than major difference.
#3. Tree shape
## CM has opaque children

Sub-nodes are just (pre)sealed models that are added to a collection. These nodes can be totally independent of the parent node. Each sub-node is an independent model that was itself sealed. The parent acts as a container for opaque children.  Child instances (with different shapes) can be changed during the lifetime of the parent object. This is useful for orders where and order consists of a number of trade legs that could be of different trade types. 
## ComposerJS parent defines shape of children

The Composer design defines the shape of the entire node tree at seal time and lists must contain items of the same shape. This is overly restrictive.

**Comment by dchambers** [comment]
Hi @james-shaw-turner,

> The current example only shows a chain starting with an external set(name, value) method. It is not clear how it works when handlers are the source of updates.

I've now [added a new section](https://github.com/dchambers/composerjs#externally-updated-handlers) covering this.

> It must then recursively call any other dependent handlers - ensuring all handlers are executed in the correct order. 

It _iteratively_ calls the handlers, and the entire code to do it, including ensuring they are executed in the correct order, is [just 5 lines](https://github.com/dchambers/composerjs/issues/5).

**Comment by james-shaw-turner** [comment]
I don't see how your "five lines" proposal could work. Imagine two handlers and three properties:
handler A listens to P1 updates P2
handler B listens to P2 updates P3
Visualised in the following DAG: P1->P2->P3
You must ensure handler-A executes before handler-B.

There appear to be two ways to do this
1. Execute all handlers in multiple passes until all updates are exhausted and have been passed down the dependency chain. The overhead for this increases exponentially dependent on the length of the chain. If any handler behaviour is expensive it must be coded to check that input parameters 
have changed.
2. Preprocess the chain - build a DAG and execute the handlers in the correct order. This also allows the detection of circular dependencies upfront. This was the approach used for derivations in the old trade model.

**Comment by yasserf** [comment]
One aspect of composing a model is to allow "business" requirements to be reusable.

So, for example we have a form with a startDate, endDate, country of departure and country of  destination.

Our conditions are:
1) I can start the form up prepopulated if required.
2) I can reset my all the fields without having to _restart_ the component.
3) I can derive how long I will be in the country of origin.

How would 1/2 be carried out in CJS without complicating the class that does the wiring/factory class (making the code less reusable)?

A code comparison for factory class:

CM:

```
model.addHandler(new ErrorHandler(model))
model.addHandler(new JourneyHandler(model))
model.addHandler(new TimeAbroadHandler(model))
model.addHandler(new DataHandler(model, "StartDate"))
model.addHandler(new DataHandler(model, "EndDate"))
model.addHandler(new DateSequenceHandler(model, ["StartDate", "EndDate"]))
```

CJS:

```
function create() {
model.setValue("StartDate", "")
model.setValue("StartDate", "EndDate")
model.setValue("Errors", "[]")
model.addHandler(["StartDate", "EndDate"], ["Errors"], validateDates)
model.addHandler(["Arrival", "Destination"], ["TimeAbroad"], calculateTimeAbroad)
model.on("reset", this, function() { model.setValue("StartDate", null), model.setValue("EndDate", null), others..  }
}

function init(data) {
  if(data.Arrival) model.setValue("Arrival", data.Arrival)
  etc..
}
```

Function vs Object:

If previous value was [1] and new value is [1,2] and my objective is to send out a message that a number has been added.

Object (in handler) / Function (added in factory method):

model.on("value", this, function(newValue, oldValue) { if(oldValue.length > newValue.length) { model.trigger("ValueAdded")} }

While circular dependencies are certainly to be frowned on, there are instances where it can't really be avoided.

| Bank (Fixed) | Spread (Can change) | Dealt (can change) |
| --- | --- | --- |
| SpotRate | SpotSpread | SpotDealt |
| ForwardPoints | ForwardSpread | ForwardDealt |
| OutrightRate | OutrightSpread | OutrightDealt |

Where if I change spot spread -> spotdealt changes -> outright dealt changes -> outright spread changes -> spotSpread changes (spot spread didn't actually change so it exits loop).

Now that should be picked up as a circular dependency in theory as if the value derived is not the same as what was entered it would end up looping forever (which has occurred before due to developer error).

**Comment by dchambers** [comment]
Ah, very good point @james-shaw-turner!

But if I understand your solution 2 correctly, it need not be any more complicated than my _five-line_ solution, as it's only when the model is being sealed that a temporary DAG is required. For example, given that all handlers are ordered within an array at seal time in an order that respects the DAG (as `this.orderedHanders`), then these six lines of code will now give us what we want, where `handlers` is a map this time instead of an array as originally proposed:

``` js
for(var handler of this.orderedHanders) {
    if(handler.id in handlers) {
        handler();
        handlers.merge(handler.dependentHandlers);
    }
}
```

Again, this is much easier to debug through than a set of listener call-backs that occur based on change throughout the node tree.

Interestingly, it's only possible to have a single list of handlers for the duration of the model due to the barring of particular nodes within node-lists being used as _input-properties_, which was necessary to satisfy the additional "No subscription management code is needed as the number of nodes within the model changes" criteria that ComposerJs adds over CM.

Oh, for reference, this is the incorrect _five-line_ version for comparison:

``` js
while(handlers.length > 0) {
    var handler = handlers.pop();
    handler();
    handlers.splice(handlers.length, 0, handler.dependentHandlers);
}
```

**Comment by dchambers** [comment]
> One aspect of composing a model is to allow "business" requirements to be reusable.
> 
> So, for example we have a form with a startDate, endDate, country of departure and country of destination.
> 
> Our conditions are:
> 1) I can start the form up prepopulated if required.
> 2) I can reset my all the fields without having to restart the component.
> 3) I can derive how long I will be in the country of origin.
> 
> How would 1/2 be carried out in CJS without complicating the class that does the wiring/factory class (making the code less reusable)?

@yasserf, I'm not sure I fully understand your questions because they seem to be exactly what CJS is designed to handle, but here are my answers anyway:
1. You can start the form-up pre-populated by invoking `set()` a number of times before sealing, or by invoking `unstringify()` before sealing. Any derived values will be calculated after you've sealed.
2. You can either invoke `set()` a number of times or use `unstringify()` to go back to a known _saved_ state. Derivations will again occur after these operations have been performed.
3. You can have a handler that derives this output-property based on a number of input-properties.

> While circular dependencies are certainly to be frowned on, there are instances where it can't really be avoided.
> 
> Bank (Fixed)  Spread (Can change) Dealt (can change)
> SpotRate  SpotSpread  SpotDealt
> ForwardPoints ForwardSpread   ForwardDealt
> OutrightRate  OutrightSpread  OutrightDealt
> Where if I change spot spread -> spotdealt changes -> outright dealt changes -> outright spread changes -> spotSpread changes (spot spread didn't actually change so it exits loop).
> 
> Now that should be picked up as a circular dependency in theory as if the value derived is not the same as what was entered it would end up looping forever (which has occurred before due to developer error).

This is an interesting requirement, and worth breaking out into a [new issue](https://github.com/dchambers/composerjs/issues/10) so it can be discussed.

**Comment by yasserf** [comment]
@dchambers 

Problem with calling stringify and unstringify is that serialise all data, including those that were derived. 

This won't work in the case of having data that is based on time, such as market prices. As such you should be able to "unstringify" only a subset of the data passed to the model to ensure everything set within the map can be handled by the model.

 My question was around where you invoke set(). If it's anywhere except the handlers you are composing your model out of it means you now have code that will have to be duplicated every time the functionality is needed.

Can you provide an example to how you would recommend two models to be used across two  or more separate classes with a mixture of handlers?

Example:

Given:
SettlementDateHandler: Adds SettlementDate and initialisation code
TenorHandler: Adds Tenor, initialisation code and translation from/to settlementDate from a service
AmountHandler: Adds Amount and initialisation code
GoodForAmountHandler: Determines if the Amount entered is greater than the maximum allowed.
UpdateHandler: Constructs a subject to request and streams prices
ServerHandler: Interacts with the server side via a DFA
FillRateHandler: Adds FillRate, initialisation code and ensures that fill rate is lower or higher than bank rate depending on model state.

RFQ Model Constructor:

```
model = new Model()
model.addHandler(new TenorHandler(model))
model.addHandler(new SettlementDateHandler(model))
model.addHandler(new AmountHandler(model))
model.addHandler(new ServerHandler(model))
model.init({Amount:500, SettlementDate:01012015, Tenor:1M})
```

Simple Tile Model Constructor:

```
model = new Model()
model.addHandler(new TenorHandler(model))
model.addHandler(new SettlementDateHandler(model))
model.addHandler(new AmountHandler(model))
model.addHandler(new UpdateHandler(model))
model.addHandler(new GoodForAmountHandler(model))
model.init({Amount:500, SettlementDate:01012015, Tenor:1M})
```

Basic Order Model Constructor:

```
model = new Model()
model.addHandler(new AmountHandler(model))
model.addHandler(new GoodForActivationDateHandler(model))
model.addHandler(new GoodForDeactivationDateHandler(model))
model.addHandler(new UpdateHandler(model))
model.addHandler(new FillRateHandler(model))
model.init({Amount:500, ActivationDate:01012015, FillRate:1})
```

Market Rate Order Model Constructor:

```
model = new Model()
model.addHandler(new AmountHandler(model))
model.addHandler(new GoodForActivationDateHandler(model))
model.addHandler(new GoodForDeactivationDateHandler(model))
model.addHandler(new UpdateHandler(model))
model.init({Amount:500, ActivationDate:01012015})
```

How would you do this is CJS with the minimal amount of lines (given each model is completely is a different class and not using inheritance)?

**Comment by dchambers** [comment]
Hi @yasserf,

I'm not going to be able to address all of your points till later, but regarding `stringify()` and `unstringify()`, I put no thought into these, and just copied them wholesale from CM, so this is something that will need to be fixed there too.

Will get back to you with more comments in due course...

**Comment by yasserf** [comment]
@dchambers 

Sounds good. One thing worth mentioning is that CM doesn't do anything other than retrieve the map on stringify (unless you pass a map of data that you want to extract and it gets the intersection). My point was the init method allows each handler to figure out what data it needs from the initialisation map to set things to the desired state and ignores other data.

Also, another small advantage to take into account is that handlers provide fixed APIs.

If I want to add an error I can do this:
    model.errors.addError("EMPTY_AMOUNT");
Rather than:
    model.set("Errors", model.getErrors().push("EMPTY_AMOUNT");
Although you could also do that statically in CJ via:
    Errors.addError(model, "EMPTY_AMOUNT") 

It would finally be good to also have an example how you respond on a field changing on a child model (or node)

theFunction = function(newValue, oldValue) {
              //do stuff
     };

model.on("CHILD_MODEL_ADDED", this, function(childModel) {
     childModel.on(fieldToListenTo, this, theFunction)
})

model.on("CHILD_MODEL_REMOVED", this, function(childModel) {
     childModel.off(fieldToListenTo, this, theFunction)
})

While also ensuring that that logic can be used in every place that has children models.
Would be nice to see an example of this using the NodeList mentioned and reusable across multiple different models.

**Comment by dchambers** [comment]
Hi @james-shaw-turner,

I hadn't originally noticed that you'd expanded the text within this issue. Since a lot of the statements are either no longer true, or were never true, these are worth addressing.

> ComposerJS handlers are called with input and output properties that it updates. ComposerJs is responsible for calling dependent handlers with the input values derived from previous handlers. CM handlers are passed a model sandbox and registers event listeners. It sets field values directly - that trigger cascaded events.

This is true

> ComposerJS has handler functions, CM has handler objects.

This is not entirely true. Firstly, the handler may be an object that exposes a function. Secondly, functions are objects in JavaScript anyway.

> ComposerJS fixes the shape of the entire object tree including sub-nodes on sealing. CM only fixes the shape of the fields of each node.

This is no longer true since [type specialization](https://github.com/dchambers/composerjs#specialized-types) is now possible.

> In both proposals a handler may be dependent on a field/property that is updated by another handler. In the general case property chaining is represented by a Directed Acyclic Graph (DAG). Both proposals must cope with any arbitrary DAG.

This is true.

> Composable Model (CM) approach
> 
> The CM registers handler objects. The handler is passed a sandboxed instance of the model in its init() method. The handler then adds listeners to any field it wants to use as an input and calls set(name, value) to update its output fields. When a property is set externally or internally (by a handler) any interested handler listeners methods are invoked - chaining follows automatically.

This is true.

> ComposerJS approach
> 
> ComposerJS registers handler functions. The function is passed model values as input and output parameters. The handler function is called whenever one of the inputs changes.
> Whenever a property is updated externally the model must find any dependent handlers and execute those. It must then recursively call any other dependent handlers - ensuring all handlers are executed in the correct order. 

As discussed in a previous comment this is not true since ComposerJs does not call handlers recursively (it calls them iteratively) and it does not need to ensure they are called in the correct order (since handler order is established when the model is sealed).

> The current example only shows a chain starting with an external set(name, value) method. It is not clear how it works when handlers are the source of updates.

The [external model usage](https://github.com/dchambers/composerjs#external-model-usage) section shows how this works.

> CM has handler objects
> 
> CM handlers are objects that implement the handler interface. As such they have two major life-cycle methods: init() which is called immediately after sealing, and destory(). This allows a handler to create any resources based on the fixed initial values of the model. e.g. a price rate provider must create a subscription that depends on currency-pair and amount. Similarly when the model is disposed of the destroy() method allows the subscription to be freed. 
> Handlers may have "business semantic" methods that can be executed directly on the handler or exported onto the model object. e.g. model.executeTrade("buy");

True.

> ComposerJS has handler functions
> 
> How does ComposerJS copes the scenarios described above.

As mentioned previously:
1. A handler object can contain methods that the model might want to invoke, in addition to the handler function.
2. The handler function can itself contain methods, like the `dispose()` method that can be used to perform resource re-allocation.

> CM has opaque children
> 
> Sub-nodes are just (pre)sealed models that are added to a collection. These nodes can be totally independent of the parent node. Each sub-node is an independent model that was itself sealed. The parent acts as a container for opaque children. Child instances (with different shapes) can be changed during the lifetime of the parent object. This is useful for orders where and order consists of a number of trade legs that could be of different trade types.

True

> ComposerJS parent defines shape of children
> 
> The Composer design defines the shape of the entire node tree at seal time and lists must contain items of the same shape. This is overly restrictive.

This is not true. ComposerJs allows children to have any shape they want. The only real difference is that sealing happens on the entire model at the same time, and this is almost certainly a good thing.

**Comment by dchambers** [comment]
Hi @yasserf, to come back to the original points I only started responding to earlier:

> My question was around where you invoke set(). If it's anywhere except the handlers you are composing your model out of it means you now have code that will have to be duplicated every time the functionality is needed.

I actually don't understand the point you are making here, but I can tell you that `set()` is only ever used within the main model code, and never by the composed model handlers.

> Can you provide an example to how you would recommend two models to be used across two or more separate classes with a mixture of handlers?

Sure, I'll replicate one of the examples you provide, but using CJS:

RFQ Model Constructor:

```
model = commonjs.create();
model.addHandler(tenorHandler);
model.addHandler(settlementDateHandler);
model.addHandler(amountHandler);
model.addHandler(serverHandler);
model.set('Amount', 500);
model.set('SettlementDate', 01012015);
model.set('Tenor', 1);
```

If you haven't already done so, please read the [main page](https://github.com/dchambers/composerjs) for this project rather than @james-shaw-turner's interpretation of the differences between CM and CJS, as I believe this will give you a much better idea as to how it works and how it's meant to be used.

**Comment by dchambers** [comment]
Hi @yasserf, more responses to your most recent comment:

> Sounds good. One thing worth mentioning is that CM doesn't do anything other than retrieve the map on stringify (unless you pass a map of data that you want to extract and it gets the intersection). My point was the init method allows each handler to figure out what data it needs from the initialisation map to set things to the desired state and ignores other data.

So are you saying that CM only serializes the same data the was passed within init()?

> It would finally be good to also have an example how you respond on a field changing on a child model (or node)

Yes, completely agree that some represenative examples of how different parts of the tree listen to each other is critical to knowing whether CJS is actually viable or not. I'm going to be looking at progressing that further tomorrow.

> While also ensuring that that logic can be used in every place that has children models.
> Would be nice to see an example of this using the NodeList mentioned and reusable across multiple different models.

Again, yes, we need lots of credible examples that mirror real world usage before we can determine whether CJS is a good idea or not.

**Comment by yasserf** [comment]
@dchambers 

> So are you saying that CM only serializes the same data the was passed within init()?

No, I meant that initialisation responsibility is distributed across the handler objects via the init method rather then via "mode.set(field, value)" in the model constructor (which causes duplication). I think @janhancic could show you a few current handlers that do this.

> If you haven't already done so, please read the main page for this project rather than @james-shaw-turner's interpretation of the differences between CM and CJS, as I believe this will give you a much better idea as to how it works and how it's meant to be used.

I have read that page initially (I believe it is getting updated though)

Two things I would like to mention:
1. The WebsocketHandler 
- Calling reexecute should happen after seal() and not on addHandler.
- An initialisation method would also useful to start the connection, rather than it happening when you add the WebsocketHandler.
1. RateHandler:
   
   ```
   var handler = function(in, out) {
   if(subscription.updateSubject('/FX/' + in.currencyPair + '/' + in.tenor + '/' + in.baseCurrency + '/' + in.amount)) {
     return false;
   }
   else {
     out.rate = subscription.data.rate;
   }
   };
   ```
- This highlights the issue that we had with the implementation of the previous DerivationApproach, which is that you don't know what changed specifically because only a single callback.

In CM the handler would instead:

```
init function() {
   //ideally we would be able to group events together using the same approach as backbone to
   //capture atomic batch updates better.
   // model.on("change:amount change:tenor change:currencyPair", this, updateSubject)

    //current method
    model.on("change:amount, this, updateSubject)
    model.on("change:tenor, this, updateSubject)
    model.on("change:currencyPair, this, updateSubject)

    subscription.on("update", this, updateData)
}
```
1. How do you propose methods can be called on handlers? 
2. Why use model.p("fieldName").on("change") instead of the model.on("change:fieldName") ? I understand the usefulness of p("fieldNameInModel").is("actuallyThisFieldInHandler") which works well in cases of using really generic handlers (as in your examples).

CM resolved this by doing:

```
addHandler(new AmountHandler(model, "NearAmount", "NEAR_AMOUNT_EMPTY_ERROR")
addHandler(new AmountHandler(model, "FarAmount")
```

Which within the AmountHandler is mapped to a variable:

```
AmountHandler(model, fieldName, emptyErrorName) {
    this.fieldName = fieldName;
    this.emptyErrorName = emptyErrorName;
}
```

This means when we debug, we use the name that we actually expect to see in the model rather than an alias.
1. Debugging:

Since all our debugging occurs in handlers (as a model is just a composition of them) will we have a way to access the state of the entire model or will we always be restricted to only see our inputs for the function?
1. Event emitting:

Since we don't have a reference to the model, how will we be able to emit an event from a handler to the outside world?
1. NodeList event changes

The NodeList mentioned is an array with the following methods:

```
length() (the number of nodes within the node-list)
item(index) (retrieve the node at the given index)
addNode(index) (add a node, optionally at a given index)
removeNode(index) (remove the node at the given index)
```

If I wanted to store a collection as a map and don't care about positions, does this mean I will have to write a wrapper around the NodeList to represent that?

Also, how many cases are there to implement a static way of having to create a tree and access properties via NodeLists? 

Following code is an example that adding nodes / removing nodes and custom behaviour can be done via custom events triggered by the external world, and the responses will then be handled appropriately by the external world (while being able to distinguish between different sets of nodes).

The Main Model: 

```
model = new Model()
model.addHandler(new ChildrenHandler(model, "TEENS")
model.addHandler(new ChildrenHandler(model, "TODDLERS")
```

A children collection:

```
function ChildrenHandler(model, type) {
   this.model = model;
   this.type = type;
   this.children = {};
}

function on(event, context, f) {
   this.model.on(event + ":" + this.type, context, f);
}

function trigger() {
   var event = arguments.splice(1);
   this.model.trigger.apply(event + ":" + this.type, arguments);
}

ChildrenHandler.init = function(data) {
  this.model.on("add", this, addChild)
  this.model.on("remove", this, removeChild)
  this.model.on("rename", this, renameChild)
  this.model.on("pause", this, pauseChild)

  createChildren(data)
}

function addChild(name, type) {
     this.children[name] =  new Child(name, type);
     this.model.trigger("added", name, new Date());
}

function removeChild(name) {
     delete this.children[name];
     this.model.trigger("removed, name, new Date());
}

function renameChild(oldName, newName) {
     this.children[newName] =  this.children[oldName];
     delete this.children[oldName];
     this.model.trigger("renamed", oldName, newName);
}

function pauseChild(name, duration) {
     this.children[name].pause(duration, resumeChild);
     this.model.trigger("paused", name, duration);
}

function resumeChild(name, duration) {
     this.model.trigger("resumed", name);
}
```

**Comment by dchambers** [comment]
> 1. The WebsocketHandler 
> - Calling reexecute should happen after seal() and not on addHandler.
> - An initialisation method would also useful to start the connection, rather than it happening when you add the WebsocketHandler.

This code was actually completely non-sensical, and has since been updated. That being said, although the handler functions are added before sealing (using `addHandler()`), they would not be invoked till after sealing.

> 1. RateHandler:
>    
>    ```
>    var handler = function(in, out) {
>    if(subscription.updateSubject('/FX/' + in.currencyPair + '/' + in.tenor + '/' + in.baseCurrency + '/' + in.amount)) {
>      return false;
>    }
>    else {
>      out.rate = subscription.data.rate;
>    }
>    };
>    ```

It's worth mentioning again that this code was non-sensical, and has since been re-written.

> - This highlights the issue that we had with the implementation of the previous DerivationApproach, which is that you don't know what changed specifically because only a single callback.
> 
> In CM the handler would instead:
> 
> ```
> init function() {
>    //ideally we would be able to group events together using the same approach as backbone to
>    //capture atomic batch updates better.
>    // model.on("change:amount change:tenor change:currencyPair", this, updateSubject)
> 
>     //current method
>     model.on("change:amount, this, updateSubject)
>     model.on("change:tenor, this, updateSubject)
>     model.on("change:currencyPair, this, updateSubject)
> 
>     subscription.on("update", this, updateData)
> }
> ```

Do you want to know which values have changed for performance reasons, or is that some handlers need to produce need different output given identical inputs, depending on which of those inputs have changed?

> 1. How do you propose methods can be called on handlers? 

I can forsee there being two ways of doing this:
1. using `composer.create()` when you want to keep methods like `set()` as private facets of the model you are building.
2. using `composer.mixinTo()` when you are happy to expose methods like `set()`.

Here's some example code using the second approach:

``` js
var composerjs = require('composerjs');
var WebSocketHandler = require('./WebSocketHandler');

function MyModel() {
    this._webSocketHandler = new WebSocketHandler('ws://acme.com/');
    composerjs.mixinTo(this);
    this.addHandler(this._webSocketHandler.handler);
    this.seal();
}

MyModel.prototype.sleepyTime = function() {
    this._webSocketHandler.pause();
    // do sleepy stuff...

    setTimeout(function() {
        this._webSocketHandler.unpause();
    }.bind(this), 3000);
};
```

Admittedly, I've still yet to add documentation for the `composer.mixinTo()` method.

> 1. Why use model.p("fieldName").on("change") instead of the model.on("change:fieldName") ? I understand the usefulness of p("fieldNameInModel").is("actuallyThisFieldInHandler") which works well in cases of using really generic handlers (as in your examples).

This is done because of the "No subscription management code is needed as the number of nodes within the model changes" goal that CJS adds over CM. Because CJS doesn't allow you to listen to particular instances of node-lists for example, fields must be described abstractly, for example `model.nodes.on('change', ...)` can be used to listen to changes on nodes that have not even been created yet.

> CM resolved this by doing:
> 
> ```
> addHandler(new AmountHandler(model, "NearAmount", "NEAR_AMOUNT_EMPTY_ERROR")
> addHandler(new AmountHandler(model, "FarAmount")
> ```
> 
> Which within the AmountHandler is mapped to a variable:
> 
> ```
> AmountHandler(model, fieldName, emptyErrorName) {
>     this.fieldName = fieldName;
>     this.emptyErrorName = emptyErrorName;
> }
> ```
> 
> This means when we debug, we use the name that we actually expect to see in the model rather than an alias.

I think you've now switched to another topic since you are now talking about handlers, whereas code like `model.nodes.on('change', ...)` is part of the public model API. By default the model will use the same field names as used internally by the handler, and where this is inappropriate this can be changed.

This is very similar to functions and methods we create as programmers, where the function uses logical easy to understand parameter names related to what that function is doing, but the names used to provide the arguments pushed in are often related to the problem domain at hand.

> 1. Debugging:
> 
> Since all our debugging occurs in handlers (as a model is just a composition of them) will we have a way to access the state of the entire model or will we always be restricted to only see our inputs for the function?

Yes, the global model code will be able to use the `get()` method for example. It's only the re-usable composed handlers that limit themselves to deriving output-properties from a set of input-properties, and they are all the simpler for it.

> 1. Event emitting:
> 
> Since we don't have a reference to the model, how will we be able to emit an event from a handler to the outside world?

Can you provide a use-case for why a handler would want to emit events without the permission of the composed model? My take on this would be that if the composed model wants to emit a message based on the output-properties that a handler has generated, then it should listen for updates, and emit events as necessary.

> 1. NodeList event changes
> 
> The NodeList mentioned is an array with the following methods:
> 
> ```
> length() (the number of nodes within the node-list)
> item(index) (retrieve the node at the given index)
> addNode(index) (add a node, optionally at a given index)
> removeNode(index) (remove the node at the given index)
> ```
> 
> If I wanted to store a collection as a map and don't care about positions, does this mean I will have to write a wrapper around the NodeList to represent that?

If you don't care about position than you won't mind using a node-list. Usually unordered collections are only provided if they confer performance benefits over using an ordered collection.

> Also, how many cases are there to implement a static way of having to create a tree and access properties via NodeLists? 
> 
> Following code is an example that adding nodes / removing nodes and custom behaviour can be done via custom events triggered by the external world, and the responses will then be handled appropriately by the external world (while being able to distinguish between different sets of nodes).

Unfortunately I don't understand the preceding paragraphs, or the point of the examples code you then go on to show.

I get the feeling you are saying that there are other ways of implementing composable models, but I think that misses the point. CJS is about having a simpler API, simpler handlers, and less bugs due to all the complexeties we've always seen when we have models that change shape over time, that involve resource management code as that happens, and that require values from servers and updates from users, all at the same time.

**Comment by yasserf** [comment]
> Do you want to know which values have changed for performance reasons, or is that some handlers need to produce need different output given identical inputs, depending on which of those inputs have changed?

I meant that we should separate different actions into different functions, not:

if(updateSubject()) {
   //update my subscription 
} else processServerUpdate(). {
   //subscription didn't change, so process a server update
}

> Yes, the global model code will be able to use the get() method for example. It's only the re-usable composed handlers that limit themselves to deriving output-properties from a set of input-properties, and they are all the simpler for it.

This moves away from the logic that all the global model code does is compose a model using only handlers and nothing else.

> Can you provide a use-case for why a handler would want to emit events without the permission of the composed model? My take on this would be that if the composed model wants to emit a message based on the output-properties that a handler has generated, then it should listen for updates, and emit events as necessary.

The code I showed in the previous example for the ChildrenClass demonstrated why a certain handler would want to emit an event.

> If you don't care about position than you won't mind using a node-list. Usually unordered collections are only provided if they confer performance benefits over using an ordered collection.

So if a user wanted to get a node by key rather than index, how would he do that?

> I get the feeling you are saying that there are other ways of implementing composable models, but I think that misses the point. CJS is about having a simpler API, simpler handlers, and less bugs due to all the complexeties we've always seen when we have models that change shape over time, that involve resource management code as that happens, and that require values from servers and updates from users, all at the same time.

My argument is more for why do we need such an advanced API? I disagree that it's simpler. CM is meant to only be an observable map with the ability to add classes with a Handler interface to interact with it. We have proven that infinite order trees of different types only need a few lines of code in the model, as well as legs, panels and other collections of data that can be changed. It's main goal is also to reduce the amounts of lines of code needed to be duplicated. I agree CJs has the same goal but it appears a large amount of lines still have to be put within the "global model" to wire things up.

**Comment by james-shaw-turner** [comment]
@dchambers  stated:

> Because CJS doesn't allow you to listen to particular instances of node-lists 

This is overly restrictive. In a swap trade there can be relationships between the near and far leg. e.g if it is locked then the near and far amounts must be the same. I'm sure there are plenty of other examples. The general rule must be you "can" listen to, or update any arbitrary field of any arbitrary node within the tree structure.

**Comment by dchambers** [comment]
Hi @yasserf, comments in-line as normal.

> I meant that we should separate different actions into different functions, not:
> 
> if(updateSubject()) {
>    //update my subscription 
> } else processServerUpdate(). {
>    //subscription didn't change, so process a server update
> }

This is already possible with CJS.

> This moves away from the logic that all the global model code does is compose a model using only handlers and nothing else.

I think the way CM _exports_ methods is wrong, for the following reasons:
1. Handlers that rely on methods for inter-connection aren't compsable.
2. Handlers that insist on exporting all of the methods they provide to the external model API are overly prescriptive.

Therefore, I think it's good that model owners decide for themselves which of these methods they want to expose.

> So if a user wanted to get a node by key rather than index, how would he do that?

They would use `addNode()` instead of `addNodeList()` when setting their model up.

> My argument is more for why do we need such an advanced API? I disagree that it's simpler. CM is meant to only be an observable map with the ability to add classes with a Handler interface to interact with it. We have proven that infinite order trees of different types only need a few lines of code in the model, as well as legs, panels and other collections of data that can be changed. It's main goal is also to reduce the amounts of lines of code needed to be duplicated.

What's simpler for you to develop as a library isn't necessarily simpler for end-developers if:
1. It requires them to write more code to do the same amount.
2. If the code they have to write is more stateful, and therefore more complex.
3. If the API is larger and harder to load into your head.

> I agree CJs has the same goal but it appears a large amount of lines still have to be put within the "global model" to wire things up.

Please take a look at the latest code examples, and you will find that CJS is simpler to work with, and requires far fewer lines than CM.

**Comment by dchambers** [comment]
Hi @james-shaw-turner,

> > Because CJS doesn't allow you to listen to particular instances of node-lists for example, fields must be described abstractly, for example model.nodes.on('change', ...) can be used to listen to changes on nodes that have not even been created yet.
> 
> This is overly restrictive. In a swap trade there can be relationships between the near and far leg. e.g if it is locked then the near and far amounts must be the same. I'm sure there are plenty of other examples. The general rule must be you "can" listen to, or update any arbitrary field of any arbitrary node within the tree structure.

The way this works is that you can't listen to a particuluar instance only, and you instead have to listen to all instances and then choose to act or not act within code instead. This is why I've been saying that the `allowMultiplexing()` method is critical to CJS, whereas we decided that it's analog in CM could be left out until we were sure we needed it.

I've now created [fx-tile](https://github.com/dchambers/composerjs/tree/master/code-examples/fx-tile) and [fx-ticket](https://github.com/dchambers/composerjs/tree/master/code-examples/fx-ticket) demos, but I'd next like to create a demo where this kind of non-symmetric processing is occurring.

**Comment by james-shaw-turner** [comment]
@yasserf 

> CM is meant to only be an observable map with the ability to add classes with a Handler interface to interact with it

There is a tension between allowing flexibility in writing handlers and increasing re-usability.  The more a component conforms to a strict interface the more reusable it is. We want people who are assembling models to have to know as little as possible about the innards of a handler. For example we need to make sure that the name of an input field in one handler matches the output field of another. Also that we don't get two handlers trying to stomp over the same field. We therefore make each handler define its input and output fields/properties. At seal time the model can do some basic checking and alert the developer of potential incompatibilities between handlers. Dom's proposal takes this a stage further by only allowing handlers access to fields that it has declared it is interested in. Yes it  makes the design appear more complex in one way, however it removes a source of errors.
The tricky thing in this design is to get the correct balance between constraining handlers to improve reusability and remove sources of error against the necessary flexibilty you describe. 

Also to say: I'm liking this compare and contrast two design approaches. It's really helping to clarify the issues so we really understand the trade-offs. So thanks for taking part :-)

**Comment by james-shaw-turner** [comment]
@dchambers  gotcha thanks for the clarification

> The way this works is that you can't listen to a particuluar instance only, and you instead have to listen to all instances and then choose to act or not act within code instead.

**Comment by dchambers** [comment]
> I've now created fx-tile and fx-ticket demos, but I'd next like to create a demo where this kind of non-symmetric processing is occurring.

@james-shaw-turner, I've now added a couple of [non-symmetric handler code examples](https://github.com/dchambers/composerjs/tree/master/code-examples/non-symmetric-handlers) to show that they can be implemented.

**Comment by dchambers** [comment]
Closing as this discussion is now coming to an end as consensus has broadly been reached.
